### PR TITLE
Publisher failures processing

### PR DESCRIPTION
**Pull request by shavkunov**
At the moment, I've added only logging.
If there is a need to resend the message, I can also do that.

Do we need slf4j or more complicated approach to logger?

Do I need to pass some kind of callback to `TaskReceiver` for external failure processing?
In that case, I just don't know, what useful that thing can do besides logging.
Logging is the thing, that I'm already doing for processing.

### CODE REVIEW

**Comment by dbarashev** [comment]
> Do we need slf4j or more complicated approach to logger?

Yes, let's use slf4j + logback. 

> Do I need to pass some kind of callback to TaskReceiver for external failure processing?

No, TaskReceiver is the one who builds and decides to publish the result, so it is his responsibility to process failures.

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt [hunk 1]
```diff
@@ -22,16 +22,19 @@ import com.google.cloud.pubsub.v1.Publisher
...
     return CompilerFleet.CompilerFleetResult.newBuilder()
             .setTaskId(taskId)
             .setStatusCode(statusCode)
-            .setResultBytes(ByteString.copyFrom(resultBytes))
+            .setResultBytes(ByteString.copyFrom(result.toByteArray()))
             .build()
             .toByteString()
 }
 
+private val logger = LoggerFactory.getLogger("Publisher")
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt]: We usually write logger in uppercase: `LOGGER`.

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt [hunk 2]
```diff
@@ -53,11 +56,12 @@ class Publisher(private val topicName: String) {
         ApiFutures.addCallback(future, object : ApiFutureCallback<String> {
 
             override fun onFailure(throwable: Throwable) {
-                println(throwable)
+                logger.info(throwable.toString())
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt]: You'll normally want error and log not only the message but also the stack trace. In slf4j API if Throwable comes as the last argument of any log methods, it prints the stack trace ([docs](https://www.slf4j.org/faq.html#paramException))

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt [hunk 3]
```diff
@@ -59,9 +60,11 @@ abstract class CompilerFleetMessageReceiver : MessageReceiver {
     abstract fun processMessage(message: PubsubMessage)
 }
 
+private val taskReceiverLogger = Logger.getLogger("TaskReceiver")
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Use slf4j here as well? And rename to `LOGGER`

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt [hunk 4]
```diff
@@ -118,10 +121,15 @@ internal class TaskReceiver(tasksDirectory: String,
         }
 
         val md5sum = dockerProcessor.getMd5Sum(rootFile)
-        this.callback("md5 sum of root file", md5sum)
+        val statusCode = 0
+        this.onMessageProcessed("md5 sum of root file", md5sum)
+
+        val onPublishFailureCallback= {
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: nit: space before `=`

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt [hunk 5]
```diff
@@ -118,10 +121,15 @@ internal class TaskReceiver(tasksDirectory: String,
         }
 
         val md5sum = dockerProcessor.getMd5Sum(rootFile)
-        this.callback("md5 sum of root file", md5sum)
+        val statusCode = 0
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: What's the point of this constant which is always 0?

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt (hunk 1 above)
Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt]: Done

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt (hunk 2 above)
Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt]: Done

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt (hunk 3 above)
Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Done

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt (hunk 5 above)
Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: I'm using this thing in two places, I didn't want to pass 0 twice.
Next time, when I'll be able to get non-zero status code it seems, that I'll need same code in logger and getResultData

I can change, if you want :)

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt (hunk 4 above)
Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Done

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt [hunk 6]
```diff
@@ -53,11 +56,12 @@ class Publisher(private val topicName: String) {
         ApiFutures.addCallback(future, object : ApiFutureCallback<String> {
 
             override fun onFailure(throwable: Throwable) {
-                println(throwable)
+                LOGGER.error("Failed on publish with message:", throwable.message, throwable)
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt]: If you want a message to be shown, add a placeholder to the string. In SLF4J placeholder is just a pair of curly braces: 
`"with message: {}"

**Comment by dbarashev** [comment]
I am getting the following when I run tests (no idea why it works on Travis):
```
java.io.IOException: The Application Default Credentials are not available. They are available if running in Google Compute Engine. Otherwise, the environment variable GOOGLE_APPLICATION_CREDENTIALS must be defined pointing to a file defining the credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
	at com.google.auth.oauth2.DefaultCredentialsProvider.getDefaultCredentials(DefaultCredentialsProvider.java:119)
	at com.google.auth.oauth2.GoogleCredentials.getApplicationDefault(GoogleCredentials.java:127)
	at com.google.auth.oauth2.GoogleCredentials.getApplicationDefault(GoogleCredentials.java:100)
	at com.google.api.gax.core.GoogleCredentialsProvider.getCredentials(GoogleCredentialsProvider.java:53)
	at com.google.cloud.pubsub.v1.Publisher.<init>(Publisher.java:157)
	at com.google.cloud.pubsub.v1.Publisher.<init>(Publisher.java:85)
	at com.google.cloud.pubsub.v1.Publisher$Builder.build(Publisher.java:661)
	at com.bardsoftware.backend.fleet.rmarkdown.Publisher.<init>(Publisher.kt:43)
	at com.bardsoftware.backend.fleet.rmarkdown.TaskReceiver.<init>(Subscriber.kt:67)
	at com.bardsoftware.backend.fleet.rmarkdown.PubsubTest.processFileUnzipTest(PubsubTest.kt:71)
```

It seems that we need to parameterize _something_. Perhaps the most reasonable is to parameterize `TaskReceiver` with publisher instance, so that it was the real one in prod and mock or fake no-operation publisher in tests.

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt (hunk 5 above)
Comment (dbarashev) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: My major point is that it is useless as status code indicator if it is always zero, no matter what happened. 
Minor point is that if you need some constant for successful or unsuccessful status codes, create an enum with semantic names: OK, FAILURE, etc.

**Comment by shavkunov** [comment]
Actually, the same problem exists in neighbor pull request.
I tried to fix this there, but I concentrated on the task.

Maybe we want to fix this in separate third pull request?

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt (hunk 6 above)
Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt]: Done

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt (hunk 5 above)
Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Replaced with enum. 
Done

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt [hunk 7]
```diff
@@ -49,6 +49,12 @@ class SubscriberArgs(parser: ArgParser) {
     )
 }
 
+enum class StatusCode(private val code: Int) {
+    SUCCESS(0), FAILURE(1);
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Not that I am feeling strong against this, but you can just declare enum constants and use their `ordinal` property: `enum class StatusCode { SUCCESS, FAILURE }`

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt [hunk 8]
```diff
@@ -121,14 +127,13 @@ internal class TaskReceiver(tasksDirectory: String,
...
-        val statusCode = 0
         this.onMessageProcessed("md5 sum of root file", md5sum)
 
         val onPublishFailureCallback = {
-            LOGGER.info("Publish failed: taskId = $taskId, status code = $statusCode, md5 sum: $md5sum")
+            LOGGER.info("Publish failed: taskId = $taskId, status code = ${StatusCode.SUCCESS.getCode()}, md5 sum: $md5sum")
         }
 
-        val data = getResultData(taskId, statusCode, md5sum)
+        val data = getResultData(taskId, StatusCode.SUCCESS.getCode(), md5sum)
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Any sense to replace int in `getResultData` with enum?

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt [hunk 9]
```diff
@@ -121,14 +127,13 @@ internal class TaskReceiver(tasksDirectory: String,
         }
 
         val md5sum = dockerProcessor.getMd5Sum(rootFile)
-        val statusCode = 0
         this.onMessageProcessed("md5 sum of root file", md5sum)
 
         val onPublishFailureCallback = {
-            LOGGER.info("Publish failed: taskId = $taskId, status code = $statusCode, md5 sum: $md5sum")
+            LOGGER.info("Publish failed: taskId = $taskId, status code = ${StatusCode.SUCCESS.getCode()}, md5 sum: $md5sum")
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Perhaps mnemonic names (which can be printed by enum's default `toString()`) are better choice for the logs.

**Comment by dbarashev** [comment]
I am okay with fixing Publisher in a separate pull request.

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt (hunk 7 above)
Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Thought, that we need to declare code explicitly.
Done

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt (hunk 9 above)
Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Done

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt (hunk 8 above)
Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Done

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt [hunk 10]
```diff
@@ -118,10 +125,14 @@ internal class TaskReceiver(tasksDirectory: String,
         }
 
         val md5sum = dockerProcessor.getMd5Sum(rootFile)
-        this.callback("md5 sum of root file", md5sum)
+        this.onMessageProcessed("md5 sum of root file", md5sum)
+
+        val onPublishFailureCallback = {
+            LOGGER.info("Publish failed: taskId = $taskId, status code: ${StatusCode.SUCCESS}, md5 sum: $md5sum")
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: I can't help crying when looking at the failure message which reports success status code :) It appears that you just don't need it here at the moment, do you?

**Review by dbarashev — APPROVED** [review]
LGTM [1 comment]

File: src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt (hunk 10 above)
Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Apparently not :)
Is it okay, that I changed to empty callback?
When there will be a status code, the callback will be more useful.

Comment (dbarashev) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Writing a message ok. Writing a little bit schizophrenic message is a little bit not ok.

Comment (shavkunov) [reply src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt]: Yeah, this message looks too verbose.

**Pull request merged**